sig/datadog/core/deprecations.rbs

```diff
@@ -0,0 +1,14 @@
+module Datadog
+  module Core
+    module Deprecations
+      @_logged_deprecations: untyped
```

Comment:
- From `Hash.new(0)` the expected hash values are `Integer`.
- From `_logged_deprecations[key] += 1` the hash values are indeed `Integer` and the keys are `key`.

To avoid repetition, let's call the `key` type "`key`" by defining a type alias.

> Note: RBS type aliases are all lowercase, distinguishing them from Ruby types which are constants and thus start with an uppercase letter.

> Note: The apparent name clash is resolved at `.rbs` parse time, as there is no actual parsing ambiguity.

Also, from our discussion, the expected behaviour of `key` is to behave correctly as a hash key, which means it must to `#hash` and `#eql?`. Let's define an interface `_Hashing` for that.

> Note: the apparent name clash is resolved at `.rbs` parse time, as there is no actual parsing ambiguity.

> Note: `_Hashing` comes from [the RBS doc](https://github.com/ruby/rbs/blob/e5cc58296fb3080a52a1bdae1a47bc12dd41a0cf/docs/syntax.md#interface-declaration). 

This gives us:

```suggestion
      interface _Hashing
        def hash: () -> ::Integer
        def eql?: (untyped) -> bool
      end

      type key = _Hashing

      @_logged_deprecations: ::Hash[key, ::Integer]
```

Reply:
So IIUC, we don't do `@logged_deprecations: ::Hash[::Object, ::Integer]` because `Object` may not respond to the functions that make it "hashable", therefore, we define an interface to which objects used as Hash keys must conform to?

Reply:
Correct. In addition:

- a `BasicObject` child instance may actually be hashable
- an `Object` child instance may have their `hash` and `eql?` methods removed
- it gets precise and explicit as to what is actually required, both in naming (`Object` is unclear as to what's the requirement is) and in which methods (through the definition of the interface)